### DISCUSSION TITLE
Find most recent build with a specific input-str

### DISCUSSION THREAD

**Discussion by dbason** [simplesurance/baur#235] [Ideas]
Currently we are using baur to support a CI/CD workflow in a mono repo that potentially has multiple teams working in it at any one time.  There are currently around 10 services in the mono repo.  We currently just use baur to calculate if a service needs to be built, and the build process is then handed off to Jenkins to produce the artifacts and upload them to Artifactory with metadata.

Every build is stamped with a semantic version that is created using gitversion.  The version also includes the short commit sha.  A build will then trigger a deployment pipeline to an environment based on what lifecycle it is (PR, merge to master, or tag).  Because code deploys are triggered by the build, and because of the version stamping we want to build every time the code changes, even if it's to revert a previous change.  This is why we want to use the `diff` command against the most recent build.

We do a build and deploy for every PR, and every time there is a push to the PR we build what has changed and deploy that.  This is because our automation is still somewhat immature and our QA teams still want to do some manual testing.  Because of this we want to keep the baur build data for a PR separate to the builds for master.  There are a couple of reasons for this:
1. So that when teams working on a different service merge to master they're not comparing against changes from a different team's PR
2. To ensure that even if code has been built in a PR we build it again when we merge to master.  This is because we stamp the binaries with a different version based on whether it is a PR build or a master build.
To enable this we are using the `--input-str` flag and using the branch or PR name as the string.  When we first open a PR we also need something to compare it against so this is where we use the `--lookup-input-str` flag with master as the branch, so we can compare against the most recent master builds.

The current functionality we are missing is a way of finding the most recent build for a specific branch/input-str rather than the most recent based only on time.  Based on the discussions we've had the most sensible option seems to be adding an option to the `baur ls runs` command to look for a build with a specific string input.  From there we can get the build number and use the diff command.

**Comment by fho** [comment]
Thanks for the detailed explanation.

It would be awesome if `baur ls runs` would support a filter syntax that provides some freedom for what you want to query, like: 
```
baur ls runs --filter "inputs CONTAINS string:master" --filter time >= 1.1.21" --sort time-desc --limit 1
```
This is a lot of work to implement though. :-)

A not so flexible but easier to implement  solution could be to add support to `baur ls runs` for:
1. a `--limit <N>` parameter, that works the same then the SQL limit clause, it limits the number of results to N.
2. a `--has-input <input-name>`  parameter, if passed only runs are shown which have an input with the given name

To find the latest run of a task in the master branch, you could then run:
```
baur ls runs --has-input <string:master> --sort time-desc --limit 1 calculator.build
```

What do you think?


>  From there we can get the build number and use the diff command.

I would be very interested to hear how you use `baur diff inputs` in your setup.
Are you using it to figure out manually why an application was rebuild? Or do you integrated it somehow into your CI workflow?

**Comment by fho** [comment]
idea was implemented here: https://github.com/simplesurance/baur/commit/0ba69df64f9e9e47d0ac61053dcab6ec85a28d4a
